listenstore/listenstore/listenstore.py

```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
 
     def __init__(self, conf):
         self.log = logging.getLogger(__name__)
-        self.replication_factor = conf.get("replication_factor", 1)
-        self.keyspace = conf.get("cassandra_keyspace", "listenbrainz")
-        host = conf.get("cassandra_server", "localhost:9092")
+
+    def max_id(self):
+        return int(self.MAX_FUTURE_SECONDS + calendar.timegm(time.gmtime()))
+
+    def fetch_listens(self, uid, from_id=None, to_id=None, limit=None):
+        """ Setup from_id, to_id, and limit for fetching listens
+        """
+        if from_id and to_id:
+            raise ValueError("You cannot specify from_id and to_id at the same time.")
+
+        if from_id:
+            order = ORDER_ASC
+        else:
+            order = ORDER_DESC
+
+        precision = 'month'
+        if from_id is None:
+            from_id = MIN_ID
+        if to_id is None:
+            to_id = self.max_id()
+        return self.fetch_listens_from_database(uid, from_id, to_id, limit, order, precision)
+
+
+class PostgresListenStore(ListenStore):
+    def __init__(self, conf):
+        ListenStore.__init__(self, conf)
+        self.log.info('Connecting to postgresql: %s', conf['sqlalchemy_database_uri'])
+        self.engine = create_engine(conf['sqlalchemy_database_uri'], poolclass=NullPool)
+        #self.connection.execute("SET synchronous_commit TO off;")
+    
+    def convert_row(self, row):
+        return Listen(uid=row[1], timestamp=row[2], artist_msid=row[3], album_msid=row[4],
+                      recording_msid=row[5], data=ujson.loads(row[6]))
+
+    def format_dict(self, listen):
+        return { 'uid': listen.uid,
+                 'id': listen.timestamp,
+                 'artist_msid': uuid.UUID(listen.artist_msid),
+                 'album_msid': uuid.UUID(listen.album_msid) if listen.album_msid is not None else None,
+                 'recording_msid': uuid.UUID(listen.recording_msid),
+                 'json': ujson.dumps(listen.data)}
+
+    @timeout(5)
+    def insert_postgresql(self, listens):
```

Comment:
I'm really glad you thought about database timeouts here! However, this mechanism isn't the proper one in this context, since the underlying database supports them. Also, deciding what to do with timeouts is hard to do now. In this case, I would suggest the following:

1. Remove this timeout feature.
2. Add a function to keep track how long postgres spends on the query.
3. Add a configuration option that allows users to specify a threshold at which point slow queries get logged. (e.g. if 3000 milliseconds are specified, then print a log statement for any query that takes longer than threshold (3000ms) to execute. We can then keep our eyes on the logs and see when things start getting into trouble.